### PR TITLE
trivial: Add SWID metadata to the binary file

### PR DESCRIPTION
**Pull request by hughsie**
If we're going to ask BIOS vendors to do this, we might as well take our own medicine.

### CODE REVIEW

File: meson.build [hunk 1]
```diff
@@ -46,4 +46,10 @@ pkgg.generate(
 
 fwupd_efi_dep = declare_dependency()
 
+# SBOM support
+uswid = find_program('uswid', required: false)
+if not uswid.found()
+  warning('uswid is unfound, so not embedding uSWID metadata')
+endif
```
[the remark this answers is not in the document]
Comment (hughsie) [reply meson.build:53]: I only merged the bits that fwupd-efi actually uses about 15 minutes ago :)

File: meson.build [hunk 2]
```diff
@@ -47,4 +47,10 @@ pkgg.generate(
...
```
Comment (superm1) [inline meson.build:54]: I think this needs to be a harder dependency now that there are reasons that you need to detect one version from another.

Comment (superm1) [reply meson.build:54]: Alternatively if uswid isn't around you can add some compatibility code that just puts a version string that can be picked up.  Then at some collective point we can drop that when python-uswid is readily available everywhere.

**Comment by superm1** [comment]
@hughsie you still planning to revive this and keep it as optional?  I don't think any of the distros are going to start packaging uswid unless something uses it, so that would at least get the ball rolling if you add it as optional in a fwupd-efi release.

**Comment by hughsie** [comment]
>  I don't think any of the distros are going to start packaging uswid unless something uses it

Is it less insane if we used a fwupd builder.xml to create the blob?

**Comment by superm1** [comment]
> > I don't think any of the distros are going to start packaging uswid unless something uses it
> 
> Is it less insane if we used a fwupd builder.xml to create the blob?

Yup

**Comment by superm1** [comment]
Still planning to do this?  If not, we should probably close the PR.

**Comment by hughsie** [comment]
Yup, still the plan I think -- will look tomorrow

File: meson.build (hunk 2 above)
Comment (hughsie) [reply meson.build:64]: > that just puts a version string

We did that in https://github.com/fwupd/fwupd-efi/commit/771a7c8e84641bfb4fc8673d8a55ab636f6a4994

File: efi/swid.xml.in [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+  versionScheme="semver"
+  xmlns="http://standards.iso.org/iso/19770/-2/2015/schema.xsd"
+  xmlns:ds="http://www.w3.org/2000/09/xmldsig#"
+  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="http://standards.iso.org/iso/19770/-2/2015-current/schema.xsd
+  swid.xsd">
+  <Meta product="fwupd-efi"/>
+  <Meta summary="EFI helpers to install system firmware"/>
+  <Meta colloquialVersion="@source_version@"/>
+  <Entity name="Richard Hughes" regid="hughsie.com" role="maintainer tagCreator"/>
```
[the remark this answers is not in the document]
Comment (hughsie) [reply efi/swid.xml.in:14]: It's less new now; still not in Fedora or Debian but there are 1.5 IBVs planning to use it.

File: efi/meson.build [hunk 4]
```diff
@@ -197,6 +217,18 @@ o_file5 = custom_target('fwup-sbat.o',
                           '--sbat-distro-url', get_option('efi_sbat_distro_url'),
                         ])
 
+if uswid.found()
+  o_files += custom_target('fwup-sbom.o',
+                          output : 'fwup-sbom.o',
+                          command : [
+                            'uswid',
+                            '--cc', cc.cmd_array()[-1],
```
Comment (superm1) [inline efi/meson.build:225]: I think this is going to fail with `ccache gcc` set to `CC`, no?

File: efi/meson.build [hunk 5]
```diff
@@ -1,6 +1,25 @@
 generate_sbat = find_program('generate_sbat.py', native: true)
 generate_binary = find_program('generate_binary.py', native: true)
 
+# get source version, falling back to empty
+git = find_program('git', required : false)
+if git.found()
+  source_version = run_command(git, 'describe').stdout().strip()
```
Comment (superm1) [inline efi/meson.build:7]: I think this needs more guarding.  What happens if you're working on a tarball in an environment with git installed?  I think this command will fail and will cause the build to fail too, no?

File: efi/meson.build [hunk 6]
```diff
@@ -1,6 +1,25 @@
...
+else
+  source_version = ''
```
Comment (superm1) [inline efi/meson.build:9]: as this ends up in the xml, perhaps this is best to be something like "unknown" rather than an empty string?

**Comment by superm1** [comment]
I think we want to make sure that a CI job is actually using this too.

File: efi/meson.build (hunk 6 above)
Comment (hughsie) [reply efi/meson.build:9]: Agreed, fixed.

File: efi/meson.build (hunk 5 above)
Comment (hughsie) [reply efi/meson.build:7]: Seems to work okay, using `gitXXX` as a test:

    Program gitXXX found: NO
    Configuring swid.xml using configuration
    ...
    [8/8] Generating efi/fwupdx64.efi with a custom command

**Comment by hughsie** [comment]
> I think we want to make sure that a CI job is actually using this too.

Agree; will add it to fwupd dependencies.xml when https://bodhi.fedoraproject.org/updates/FEDORA-2023-d26d3444bf hits stable in 7 days.

File: efi/meson.build (hunk 5 above)
Comment (superm1) [reply efi/meson.build:7]: I meant having git program but no .git directory. Did you check that?

Comment (hughsie) [reply efi/meson.build:7]: Ahh! Gotcha, sorry. It does, but only by accident. I've added a `check: false` to be explicit about it.

**Review by superm1 — APPROVED** [review]

**Comment by hughsie** [comment]
I'll merge the python3-uswid dep into contrib/ci/dependencies.xml when the package hits fedora. Thanks for all the feedback on this PR.

**Pull request merged**